src/papi_preset.c

```diff
@@ -1290,6 +1330,457 @@ papi_load_derived_events (char *pmu_str, int pmu_type, int cidx, int preset_flag
+				if (derived == DERIVED_INFIX) {
+					SUBDBG( "Converting InFix operations %s\n", t);
+					t = infix_to_postfix( t );
+					results[res_idx].derived_int = DERIVED_POSTFIX;
+				}
+
+				SUBDBG( "Saving PostFix operations %s\n", t);
+				results[res_idx].postfix = papi_strdup(t);
+			}
+
+			/* All derived terms collected here */
+			i = 0;
+			invalid_event = 0;
+			results[res_idx].count = 0;
+            int firstTerm = 1;
+			do {
+				t = trim_string(strtok_r(NULL, ",", &tok_save_ptr));
+				if ((t == NULL) || (strlen(t) == 0))
+					break;
+				if (strcasecmp(t, "NOTE") == 0)
+					break;
+				if (strcasecmp(t, "LDESC") == 0)
+					break;
+				if (strcasecmp(t, "SDESC") == 0)
+					break;
+
+				SUBDBG( "Adding term (%d) %s to derived event %#x, current native event count: %d.\n", i, t, preset, results[res_idx].count);
+
+				// show that we do not have an event code yet (the component may create one and update this info)
+				// this also clears any values left over from a previous call
+				_papi_hwi_set_papi_event_code(-1, -1);
+
+                /* Do the loop that scrapes qualifiers here. */
+                unsigned int eventCode;
+                char *tmpEvent;
+                PAPI_event_info_t eventInfo;
+                hwi_presets_t *prstPtr = &(_papi_hwi_comp_presets[cidx][preset_index]);
+
+                if( firstTerm ) {
+
+                    // Convert native event to code and check that it's valid.
+                    status = _papi_hwi_native_name_to_code(t, &eventCode);
+                    if( status != PAPI_OK ) {
+					    invalid_event = 1;
+					    PAPIERROR("Failed to get code for native event %s, used in derived event %s", t, results[res_idx].symbol);
+					    break;
+                    }
+
+                    do {
+                        // Call get_event_info, and use the qualifier string that comes after the
+                        // single instance of ":" and the description that comes after "masks:"
+                        status = _papi_hwi_get_native_event_info( (unsigned int)eventCode, &eventInfo );
+                        if ( status != PAPI_OK ) {
+					        invalid_event = 1;
+					        PAPIERROR("Failed to get info for native event %s, used in derived event %s", t, results[res_idx].symbol);
+					        break;
+                        }
+
+                        const char *wholeNameConst = _papi_hwi_strip_component_prefix(eventInfo.symbol);
+                        char *wholeName = strdup(wholeNameConst);
+                        char *qualDelim = ":";
+                        char *qualPtr = strstr( wholeName, qualDelim );
+                        if( NULL != qualPtr ) {
+
+                            prstPtr->num_quals++;
+
+                            /* Store the qualifier in the preset struct. */
+                            prstPtr->quals[prstPtr->num_quals-1] = (char*)malloc((1+strlen(qualPtr))*sizeof(char));
```

Comment:
Why not increment `prstPtr->num_quals++` further down? Then you would not have to do `prstPtr->num_quals-1`. I may be missing something though and you actually need to increment at that point.